presto-main/src/main/java/com/facebook/presto/cost/CostCalculatorUsingExchanges.java

```diff
@@ -0,0 +1,254 @@
+                    node.getRight(),
+                    Objects.equals(node.getDistributionType(), Optional.of(JoinNode.DistributionType.REPLICATED)));
+        }
+
+        private PlanNodeCostEstimate calculateJoinCost(PlanNode join, PlanNode probe, PlanNode build, boolean replicated)
+        {
+            int numberOfNodesMultiplier = replicated ? numberOfNodes : 1;
+
+            PlanNodeStatsEstimate probeStats = getStats(probe);
+            PlanNodeStatsEstimate buildStats = getStats(build);
+            PlanNodeStatsEstimate outputStats = getStats(join);
+
+            double buildSideSize = buildStats.getOutputSizeInBytes(build.getOutputSymbols());
+            double probeSideSize = probeStats.getOutputSizeInBytes(probe.getOutputSymbols());
+            double joinOutputSize = outputStats.getOutputSizeInBytes(join.getOutputSymbols());
+
+            double cpuCost = probeSideSize +
+                    buildSideSize * numberOfNodesMultiplier +
+                    joinOutputSize;
+
+            if (replicated) {
+                // add the cost of a local repartitioning of build side copies
+                // cost of the repartitioning of a single data copy has been already added in calculateExchangeCost
+                cpuCost += buildSideSize * (numberOfNodesMultiplier - 1);
+            }
+
+            double memoryCost = buildSideSize * numberOfNodesMultiplier;
+
+            return new PlanNodeCostEstimate(cpuCost, memoryCost, 0);
+        }
+
+        @Override
+        public PlanNodeCostEstimate visitExchange(ExchangeNode node, Void context)
+        {
+            return calculateExchangeCost(numberOfNodes, getStats(node), node.getOutputSymbols(), node.getType(), node.getScope());
+        }
+
+        @Override
+        public PlanNodeCostEstimate visitSemiJoin(SemiJoinNode node, Void context)
+        {
+            return calculateJoinCost(
+                    node,
+                    node.getSource(),
+                    node.getFilteringSource(),
+                    node.getDistributionType().orElse(SemiJoinNode.DistributionType.PARTITIONED).equals(SemiJoinNode.DistributionType.REPLICATED));
+        }
+
+        @Override
+        public PlanNodeCostEstimate visitValues(ValuesNode node, Void context)
+        {
+            return ZERO_COST;
+        }
+
+        @Override
+        public PlanNodeCostEstimate visitEnforceSingleRow(EnforceSingleRowNode node, Void context)
+        {
+            return ZERO_COST;
+        }
+
+        @Override
+        public PlanNodeCostEstimate visitLimit(LimitNode node, Void context)
+        {
+            // This is just a wild guess. First of all, LimitNode is rather rare except as a top node of a query plan,
+            // so proper cost estimation is not that important. Second, since LimitNode can lead to incomplete evaluation
+            // of the source, true cost estimation should be implemented as a "constraint" enforced on a sub-tree and
```

Comment:
This is not trivial and would require slight redesign how we calculate cost. Before we dive into that, though, we should have some sense how often this would be helpful -- how often there is a LIMIT without ORDER BY in a query sub-tree (subquery, etc.)